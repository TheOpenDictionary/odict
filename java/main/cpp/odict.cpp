#include <jni.h>
#include <stdio.h>
#include <iostream>
#include "bridge/bridge.h"

using namespace std;

#ifndef CGO_EXPORT_BRIDGE_EXISTS
#error bridging header not found
#endif

extern "C" JNIEXPORT jstring JNICALL
Java_org_odict_Dictionary_lookup(JNIEnv *env, jobject, jstring query, jstring dictID)
{
  const char *dictionaryID = env->GetStringUTFChars(dictID, 0);
  const char *entry_term = env->GetStringUTFChars(query, 0);
  char *result = LookupEntry((char *)entry_term, (char *)dictionaryID);
  jstring str = env->NewStringUTF(result);

  free(result);

  return str;
}

extern "C" JNIEXPORT void JNICALL
Java_org_odict_Dictionary_index(JNIEnv *env, jobject, jstring path, jboolean force)
{
  char *p = (char *)env->GetStringUTFChars(path, 0);
  
  IndexDictionary(p, force);

  env->ReleaseStringUTFChars(path, p);
}

extern "C" JNIEXPORT jstring JNICALL
Java_org_odict_Dictionary_search(JNIEnv *env, jobject, jstring query, jstring dictionary_id)
{
  char *q = (char *)env->GetStringUTFChars(query, 0);
  char *dictid = (char *)env->GetStringUTFChars(dictionary_id, 0);
  const char *result = SearchDictionary(q, dictid);

  jstring res = env->NewStringUTF(result);

  env->ReleaseStringUTFChars(query, q);
  env->ReleaseStringUTFChars(dictionary_id, dictid);

  free((char *)result);

  return res;
}

extern "C" JNIEXPORT jstring JNICALL
Java_org_odict_Dictionary_read(JNIEnv *env, jobject, jstring path)
{
  const char *dictionary_path = env->GetStringUTFChars(path, 0);
  char *encoded = ReadDictionary((char *)dictionary_path);

  jstring res = env->NewStringUTF(encoded);

  env->ReleaseStringUTFChars(path, dictionary_path);

  free(encoded);

  return res;
}

extern "C" JNIEXPORT void JNICALL
Java_org_odict_Dictionary_write(JNIEnv *env, jobject, jstring xml, jstring path)
{
  const char *dictionary_path = env->GetStringUTFChars(path, 0);
  const char *content = env->GetStringUTFChars(xml, 0);

  WriteDictionary((char *)content, (char *)dictionary_path);

  env->ReleaseStringUTFChars(path, dictionary_path);
  env->ReleaseStringUTFChars(xml, content);
}

extern "C" JNIEXPORT void JNICALL
Java_org_odict_Dictionary_compile(JNIEnv *env, jobject, jstring path)
{
  const char *dictionary_path = env->GetStringUTFChars(path, 0);

  CompileDictionary((char *)dictionary_path);

  env->ReleaseStringUTFChars(path, dictionary_path);
}
