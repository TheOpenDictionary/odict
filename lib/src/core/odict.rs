/* -------------------------------------------------------------------------- */
/*                               DictionaryFile                               */
/* -------------------------------------------------------------------------- */

use std::path::PathBuf;

use rkyv::access_unchecked;

use crate::{semver::SemanticVersion, ArchivedDictionary, Dictionary};

#[derive(Clone)]
pub struct OpenDictionary {
    pub signature: String,
    pub version: SemanticVersion,
    pub path: Option<PathBuf>,
    pub total_size: u64,

    pub(super) content: Vec<u8>,
}

impl OpenDictionary {
    pub fn access(&self) -> crate::Result<&ArchivedDictionary> {
        let archived = unsafe { access_unchecked::<crate::ArchivedDictionary>(&self.content[..]) };
        Ok(archived)
    }

    pub fn deserialize(&self) -> crate::Result<Dictionary> {
        let dict: Dictionary = self.access()?.deserialize()?;
        Ok(dict)
    }
}
