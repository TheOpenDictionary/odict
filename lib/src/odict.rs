/* -------------------------------------------------------------------------- */
/*                               DictionaryFile                               */
/* -------------------------------------------------------------------------- */

use std::path::PathBuf;

use crate::version::SemanticVersion;

#[derive(Clone)]
pub struct OpenDictionary {
    pub(crate) signature: String,
    pub(crate) version: SemanticVersion,
    pub(crate) path: Option<PathBuf>,
    pub(crate) bytes: Vec<u8>,
}

impl OpenDictionary {
    pub fn signature(&self) -> &str {
        &self.signature
    }

    pub fn path(&self) -> &Option<PathBuf> {
        &self.path
    }

    pub fn version(&self) -> &SemanticVersion {
        &self.version
    }

    pub fn size(&self) -> crate::Result<u64> {
        if let Some(path) = &self.path {
            if let Ok(metadata) = std::fs::metadata(path) {
                return Ok(metadata.len());
            }
        }

        let bytes = self.to_bytes()?;

        Ok(bytes.len().try_into()?)
    }

    pub fn contents(&self) -> crate::Result<&crate::schema::ArchivedDictionary> {
        Ok(rkyv::access::<
            crate::schema::ArchivedDictionary,
            rkyv::rancor::Error,
        >(&self.bytes[..])?)
    }
}
