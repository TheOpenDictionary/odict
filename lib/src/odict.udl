namespace odict {
};

interface DictionaryFile {};

interface DictionaryReader {
  constructor();
};