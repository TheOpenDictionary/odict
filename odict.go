package main

import (
	"log"
	"os"

	"github.com/TheOpenDictionary/odict/cli"
	"github.com/TheOpenDictionary/odict/lib/validator"
)

func main() {
	validator.InitializeValidator()

	defer validator.CleanupValidator()

	err := cli.App.Run(os.Args)

	if err != nil {
		log.Fatal(err)
	}
}
