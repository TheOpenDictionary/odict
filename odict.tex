\documentclass{article}
\usepackage{hyperref}
\begin{document}
\title{\textbf{The Open Dictionary Project}}
\author{\textbf{Tyler Nickerson} \\ \small{Linguistic Inc.} \\ \small{tyler@linguistic.io}}
\date{}
\maketitle
The current state of digital dictionary technology presents numerous challenges for those seeking offline and programmatic access to lexical data. Most dictionaries still in use today are built upon outdated or proprietary technologies, making them challenging to adopt in modern applications. This lack of standardization has led to a proliferation of custom protocols for storing data across many offline dictionary applications. This problem is further compounded by the fact that lexemes are typically stored as HTML fragments, making the extraction of structured data both difficult and variable across different formats.

In response to these challenges, we propose the Open Dictionary Project (ODict) as a solution for creating a new, open standard for dictionary formats that is both extensible and user-friendly. ODict includes a public file format specification, a command-line compiler, as well as wrapper libraries for creating and accessing lexical dictionaries. Dictionaries are compiled from human-readable XML and stored in a single, compressed binary file that can be easily transmitted and accessed on any device. Lexical information, such as etymologies, senses, and definitions, is stored in a buffer as structured data that can be output in a variety of formats. This approach allows language scientists to leverage well-formed lexical data in their own projects without having to process HTML.

Lexemes can be instantly retrieved \emph{without} deserializing the file buffer and \emph{without} relying on external index files, making lookups extremely fast. Furthermore, unlike other formats, ODict features official language bindings for NodeJS, Go, Python, and JVM languages (although the ODict specification can be used in any programming language supported by its underlying FlatBuffer protocol).

In this presentation, the process for writing, compiling, and consuming ODict files will be demonstrated, along with a discussion around the motivation and development of the format as described above. We will also share our future plans for the Open Dictionary Project in order to encourage a constructive discourse among conference attendees. The development of a new, open standard for dictionary formats is an important step towards improving the accessibility and reliability of lexical data for all users. We believe ODict has great potential in the fields of natural language processing and language learning, and we hope to see the format adopted by the wider community.
\end{document}
