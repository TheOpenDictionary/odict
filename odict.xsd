	
	<xs:schema attributeFormDefault="unqualified" elementFormDefault="qualified" xmlns:xs="http://www.w3.org/2001/XMLSchema">
	  <xs:element name="dictionary">
	    <xs:complexType>
	      <xs:sequence>
	        <xs:element minOccurs="1" maxOccurs="unbounded" name="entry">
	          <xs:complexType>
	            <xs:sequence>
	              <xs:element minOccurs="1" maxOccurs="unbounded" name="ety">
	                <xs:complexType>
	                  <xs:sequence>
	                    <xs:element minOccurs="1" maxOccurs="unbounded" name="sense">
	                      <xs:complexType>
	                        <xs:sequence>
	                          <xs:element minOccurs="1" maxOccurs="unbounded" name="definition">
	                            <xs:complexType>
	                              <xs:sequence>
	                                <xs:element minOccurs="1" maxOccurs="unbounded" name="example" type="xs:string" />
	                                <xs:element minOccurs="1" maxOccurs="unbounded" name="note">
	                                  <xs:complexType>
	                                    <xs:sequence>
	                                      <xs:element minOccurs="1" maxOccurs="unbounded" name="example" type="xs:string" />
	                                    </xs:sequence>
	                                    <xs:attribute name="id" type="xs:string" />
	                                    <xs:attribute use="required" name="value" type="xs:string" />
	                                  </xs:complexType>
	                                </xs:element>
	                              </xs:sequence>
	                              <xs:attribute name="id" type="xs:string" />
	                              <xs:attribute use="required" name="value" type="xs:string" />
	                            </xs:complexType>
	                          </xs:element>
	                          <xs:element minOccurs="1" maxOccurs="unbounded" name="group">
	                            <xs:complexType>
	                              <xs:sequence>
	                                <xs:element minOccurs="1" maxOccurs="unbounded" name="definition">
	                                  <xs:complexType>
	                                    <xs:sequence>
	                                      <xs:element minOccurs="1" maxOccurs="unbounded" name="example" type="xs:string" />
	                                      <xs:element minOccurs="1" maxOccurs="unbounded" name="note">
	                                        <xs:complexType>
	                                          <xs:sequence>
	                                            <xs:element minOccurs="1" maxOccurs="unbounded" name="example" type="xs:string" />
	                                          </xs:sequence>
	                                          <xs:attribute name="id" type="xs:string" />
	                                          <xs:attribute use="required" name="value" type="xs:string" />
	                                        </xs:complexType>
	                                      </xs:element>
	                                    </xs:sequence>
	                                    <xs:attribute name="id" type="xs:string" />
	                                    <xs:attribute use="required" name="value" type="xs:string" />
	                                  </xs:complexType>
	                                </xs:element>
	                              </xs:sequence>
	                              <xs:attribute name="id" type="xs:string" />
	                              <xs:attribute use="required" name="description" type="xs:string" />
	                            </xs:complexType>
	                          </xs:element>
	                        </xs:sequence>
	                        <xs:attribute use="required" name="pos" type="xs:string" />
	                      </xs:complexType>
	                    </xs:element>
	                  </xs:sequence>
	                  <xs:attribute name="id" type="xs:string" />
	                  <xs:attribute name="description" type="xs:string" />
	                </xs:complexType>
	              </xs:element>
	            </xs:sequence>
	            <xs:attribute use="required" name="term" type="xs:string" />
	            <xs:attribute name="pronunciation" type="xs:string" />
	            <xs:attribute name="see" type="xs:string" />
	          </xs:complexType>
	        </xs:element>
	      </xs:sequence>
	      <xs:attribute use="required" name="id" type="xs:string" />
	      <xs:attribute name="name" type="xs:string" />
	    </xs:complexType>
	  </xs:element>
	</xs:schema>