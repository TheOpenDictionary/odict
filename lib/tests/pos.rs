#[cfg(test)]
mod pos_tests {

    use insta::assert_snapshot;
    use map_macro::hash_map;
    use odict::{format::xml::ToXML, Dictionary, Entry, Etymology, PartOfSpeech, Sense};

    #[test]
    fn test_display() {
        assert_eq!(
            PartOfSpeech::AdjKari.to_string(),
            "'kari' adjective (archaic)"
        );
        assert_eq!(PartOfSpeech::Other("custom".into()).to_string(), "custom");
    }

    #[test]
    fn test_se_custom() {
        let mut dict = Dictionary::default();

        dict.id = "2ee2a1ae-f7ff-4590-ba2d-de857ba7857f".try_into().unwrap();

        dict.entries.insert(
            "dog".into(),
            Entry {
                term: "dog".into(),
                see_also: None,
                etymologies: vec![Etymology {
                    id: None,
                    pronunciations: vec![],
                    description: None,
                    senses: hash_map! {
                      PartOfSpeech::Other("cusTom".into()) => Sense {
                        lemma: None,
                        definitions: vec![],
                        tags: vec![],
                        translations: vec![],
                        forms: vec![],
                        pos: PartOfSpeech::Other("cusTom".into())
                      }
                    },
                }],
            },
        );

        assert_snapshot!(dict.to_xml(true).unwrap());
    }

    #[test]
    fn test_de_custom() {
        let xml = "
          <dictionary>
            <entry term=\"dog\">
              <ety>
                <sense pos=\"custom\" />
              </ety>
            </entry>
          </dictionary>
        ";

        let dict = Dictionary::try_from(xml).unwrap();

        let expected = PartOfSpeech::Other("custom".into());

        let sense = dict.entries.get("dog").unwrap().etymologies[0]
            .senses
            .get(&expected);

        assert!(sense.is_some());
        assert_eq!(sense.unwrap().pos, expected);
    }

    #[test]
    fn test_se() {
        let mut dict = Dictionary::default();

        dict.id = "2ee2a1ae-f7ff-4590-ba2d-de857ba7857f".try_into().unwrap();

        dict.entries.insert(
            "dog".into(),
            Entry {
                term: "dog".into(),
                see_also: None,
                etymologies: vec![Etymology {
                    id: None,
                    pronunciations: vec![],
                    description: None,
                    senses: hash_map! {
                      PartOfSpeech::AdjKari => Sense {
                        lemma: None,
                        definitions: vec![],
                        tags: vec![],
                        translations: vec![],
                        forms: vec![],
                        pos: PartOfSpeech::AdjKari
                      }
                    },
                }],
            },
        );

        assert_snapshot!(dict.to_xml(true).unwrap());
    }

    #[test]
    fn test_de() {
        let xml = "
          <dictionary>
            <entry term=\"dog\">
              <ety>
                <sense pos=\"adj_kari\" />
              </ety>
            </entry>
          </dictionary>
        ";

        let dict = Dictionary::try_from(xml).unwrap();

        let sense = dict.entries.get("dog").unwrap().etymologies[0]
            .senses
            .get(&PartOfSpeech::AdjKari);

        assert!(sense.is_some());
        assert_eq!(sense.unwrap().pos, PartOfSpeech::AdjKari);
    }
}
