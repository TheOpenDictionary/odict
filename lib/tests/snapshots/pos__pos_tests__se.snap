---
source: lib/tests/pos.rs
expression: dict.to_xml(true).unwrap()
---
<dictionary id="2ee2a1ae-f7ff-4590-ba2d-de857ba7857f">
  <entry term="dog">
    <ety>
      <sense pos="adj_kari"/>
    </ety>
  </entry>
</dictionary>
